### PR TITLE
Stratum v2 Noise Protocol

### PR DESCRIPTION
**Pull request by Sjors**
Parent PR #29432. Followed by #30315.

This is the first part of the Stratum v2 Template Provider change that can be a standalone pull request.

It introduces a CMake build flag `-DWITH_SV2` that is `OFF` by default (but `ON` for CI).

The Noise Protocol Framework is defined here: https://noiseprotocol.org/noise.html

It's quite similar to BIP324. The main differences are explained here, including why Stratum v2 can't use BIP234 (yet): https://delvingbitcoin.org/t/stratum-v2-noise-protocol-bip324-nuggets/413

The implementation is based on revision 38, 2018-07-11 (most recent at the time of writing).

The Stratum v2 spec defines the specific choice of ciphers: https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md

This protocol has been implemented in the Stratum Reference Implementation (SRI, using [rust-bitcoin](https://github.com/rust-bitcoin/rust-bitcoin)). https://github.com/stratum-mining/stratum

It has also been (independently) implemented in BraiinsOS. This part is currently not open source, and it's behind on some recent changes.

The classes in `sv2_noise.h` attempt to stay close to the paper, whereas the test and fuzzer code borrow heavily from BIP324.

It's ready for review, but not for merge:

1. The parent PR may need more conceptual review (and perhaps the entire spec, but that would really slow things down)
2. We could decide to not support Noise encryption at all and require users to install separate software for that. Code review of this PR could help inform that decision.

### CODE REVIEW

**Review by winlover32 — APPROVED** [review]

**Review by winlover32 — APPROVED** [review]

File: src/common/sv2_noise.h
Comment (Sjors) [reply src/common/sv2_noise.h:76]: Any suggested reading to wrap my head around this serialisation / bytes / span stuff?

Comment (Sjors) [reply src/common/sv2_noise.h:77]: Since `m_sig` is always 64 bytes, I changed it to `std::array` (and using `std::byte` for consistency).

**Comment by Sjors** [comment]
I moved the static constants into classes and renamed them a bit. Expanded the fuzzer to send multiple messages back and forth (like the BIP324 fuzzer).

I dropped the 1 hour wiggle room in certificate timestamps, because it adds complexity and I don't expect this to cause any issues in practice. A self-signed certificate is generated when the TemplateProvider loads, which is unlikely the same second someone connects to it.

The fuzzer found a bug where I forgot to defragment messages larger than the chunk size. Since the Template Provider generally only _sends_ large messages, this (probably) wasn't caught in testing.

[it's currently in a broken state, second ECDH handshake gives different result, maybe due to bad memory management]

File: src/common/sv2_noise.h
Comment (Sjors) [reply src/common/sv2_noise.h:66]: I'll look into replace `write` with `>>`.

Not sure if it's worth changing `m_sig` from `std::array<unsigned char>` to `std::array<std::byte>`. It's only used by `VerifySchnorr` and `SignSchnorr` which want a `Span<unsigned char>`, so whether we cast it here or there doesn't matter?

Comment (Sjors) [reply src/common/sv2_noise.h:66]: ` << m_sig` works fine, so `m_sig` can live happily as a `std::array<unsigned char`

**Comment by Sjors** [comment]
Alright, code is back in working state. I dropped a bunch of spurious `Make*ByteSpan` (mostly by switching remaining uses of `uint8_t` to `std::byte`).

Also switched to the new logging convention, mostly `LogTrace()`. Also expanded the fuzzer to mess with bytes during the handshake.

**Comment by Sjors** [comment]
This latest push improves support for mock time (which the parent PR needs). It's still using the deprecated `GetTime`, but it's not clear to me what to replace it with.

It also changes the protocol name to `Noise_NX_Secp256k1+EllSwift_ChaChaPoly_SHA256` (https://github.com/stratum-mining/sv2-spec/pull/66#discussion_r1483090409).

**Comment by Sjors** [comment]
Updated and simplified the description to account for the new parent PR. I'm leaving this in draft status pending two merges on the SRI side; that way the whole thing is easier to test.

**Comment by Sjors** [comment]
Possibly relevant: the test introduced here, and not modified later, failed at least once in the parent PR: https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-1948481906

**Comment by Sjors** [comment]
(I plan to rebase this after ASan CI is fixed)

**Comment by Sjors** [comment]
Rebased after #29419.

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/pull/29346#issue-2105856059

> We could decide to not support Noise encryption at all and require users to install separate software for that. Code review of this PR could help inform that decision.

I'm just starting to look at this PR and #29432, and having separate software to handle the encryption and networking does seem like a pretty appealing approach.

However, even if the goal is to directly add support for the stratum protocol to bitcoin core, I think it would be good to have a little more separation between the stratum and node code, so instead of accessing the [mempool](https://github.com/Sjors/bitcoin/blob/5825cd80a6d67543f62757f29e0d27117209f2f4/src/node/sv2_template_provider.cpp#L274) and [blockassembler](https://github.com/Sjors/bitcoin/blob/5825cd80a6d67543f62757f29e0d27117209f2f4/src/node/sv2_template_provider.cpp#L480C26-L480C70) and [chainman](https://github.com/Sjors/bitcoin/blob/5825cd80a6d67543f62757f29e0d27117209f2f4/src/node/sv2_template_provider.cpp#L726) and [g_best_block](https://github.com/Sjors/bitcoin/blob/5825cd80a6d67543f62757f29e0d27117209f2f4/src/node/sv2_template_provider.cpp#L265) directly, the stratum code would call a more well-defined interface. This could be a new interface class in `src/interfaces/` called something like `MiningServer` or `BlockTemplateProvider` with a few methods for polling the mempool and chainstate, generating block templates and submitting new blocks.

I think this would be good in longer term to have more decoupling and potential process and project separation. But it could also help review in shorter term, by giving reviewers like me who are more familiar with bitcoin core code than with the stratum protocol another entry point for understanding the code.

**Comment by ryanofsky** [comment]
> > the stratum code would call a more well-defined interface
> 
> I would like that too. Just not sure how to go about it.

I think in #29432, you could add a `src/interfaces/mining_service.h` header similar to [`src/interfaces/chain.h`](https://github.com/bitcoin/bitcoin/blob/master/src/interfaces/chain.h) that looks something like:

```c++
class MiningService {
public:
    virtual ~MiningService() {}

    // Returns a new block template based on the current state of the mempool and blockchain
    virtual std::unique_ptr<CBlockTemplate> GetNewBlockTemplate(const CScript& scriptPubKey) = 0;

    // Accepts a completed block from a miner and attempts to add it to the blockchain
    virtual bool SubmitNewBlock(const std::shared_ptr<const CBlock>& block) = 0;

    // Returns the current best block hash
    virtual uint256 GetCurrentBestBlock() = 0;

    // Triggers an update or checks the status of the mempool
    virtual void UpdateMempool() = 0;

    // Additional methods can be added here as needed
};
```

^^^ was suggested by chatgpt after I uploaded `sv2_template_provider.cpp`

Implementation of the interface could go in `src/node/interfaces.cpp` and I think only `sv2_template_provider.cpp` would need to call the interface unless there are other parts of the PR doing similar things.

Or another approach if we wanted to move polling out of `sv2_template_provider.cpp` into the node could be an interface more like:

```c++
class MiningService {
public:
    virtual ~MiningService() {}

    // Register a callback for new block templates
    virtual std::unique_ptr<Handler> RequestBlockTemplates(const CScript& scriptPubKey, std::function<void(const CBlockTemplate&>callback)) = 0;

    // Accepts a completed block from a miner and attempts to add it to the blockchain
    virtual bool SubmitNewBlock(const std::shared_ptr<const CBlock>& block) = 0;
};
```

I don't think defining an interface is necessary, but it could be a useful way of breaking up the code.

**Comment by Sjors** [comment]
@ryanofsky thanks for the suggestion! Maybe I could start by having the `getblocktemplate` RPC use that new interface? So instead of having:

```cpp
NodeContext& node = EnsureAnyNodeContext(request.context);
ChainstateManager& chainman = EnsureChainman(node);
...
Chainstate& active_chainstate = chainman.ActiveChainstate();
...
const CTxMemPool& mempool = EnsureMemPool(node);
...
pblocktemplate = BlockAssembler{active_chainstate, &mempool}.CreateNewBlock(scriptDummy);
```

I would write something like:

```cpp
MiningContext& miner = EnsureAnyMinerContext(request.context);
...
miner.GetNewBlockTemplate(scriptDummy)
```

And then interface implementation would take care of getting the chain manager and mempool (via the node interface?).

**Comment by ryanofsky** [comment]
> Maybe I could start by having the `getblocktemplate` RPC use that new interface?

That's a good idea, nice way the interface could be used from the start.

> And then interface implementation would take care of getting the chain manager and mempool (via the node interface?).

Right, you could look at the interfaces::Chain for an example of how to initialize the interface. Store the interface in NodeContext like:

https://github.com/bitcoin/bitcoin/blob/7fcf4e99791ca5be0b068ac03a81a50ece11dba3/src/node/context.h#L71

Initialize in the init.cpp file like:

https://github.com/bitcoin/bitcoin/blob/7fcf4e99791ca5be0b068ac03a81a50ece11dba3/src/init.cpp#L1113

You could just copy the makeChain method there, which has the boilerplate used to instantiate the interface and give it access to the chainstate and mempool.

Then add an Ensure function like you mentioned in https://github.com/bitcoin/bitcoin/blob/master/src/rpc/server_util.h

I'd maybe call the interface something like MiningService or TemplateProvider rather than MiningContext. I think of "Context" as being a passive container for state instead of an object you would call methods on. The top level interfaces like `interfaces::Chain` `interface::Node` `interfaces::Wallet` are actually stateless themselves, and only point to state stored in other places. (The idea is to let interface instances be created and destroyed freely, since with multiprocess support new processes, can connect and create interface pointers that get destroyed on disconnection, and the state of the node shouldn't be affected too much from connections and disconnections. Anyway, the mining interface could follow this pattern.)

---

I am also still curious about the idea to "to not support Noise encryption at all and require users to install separate software." I'm not sure if this has already been discussed somewhere or if you have more thoughts on it. Looking at this PR, the noise protocol implementation does not seem too complicated, and seems like it would not be a big deal to review and maintain as part of bitcoin core, but I'm not sure yet about the rest of the mining protocol. I'm also not sure how much work it would be to create separate software, if there could be a middle ground for mining support like there is for gui support, which is distributed with bitcoin core, but in a separate executable and with a separate github repository.

**Comment by Sjors** [comment]
> if you have more thoughts on it

Whenever I talk to people interested in using Stratum v2 I find it difficult to explain all the moving parts they need to install and configure.

Install:
* a separate Bitcoin Core branch (temporary, pending #29432 merge)
* a job declaration client
* a sv1-sv2 translator (temporary, pending sv2 firmware support)

Configure:
* point job declaration client to Bitcoin Core, to the pools job declaration server and to the pool itself
* point translator to job declaration client
* point miner to translator
* various public keys for noise encryption (optionally private key and certificate)

Ideally mining should be as easy as: install Bitcoin Core, paste a single [pool URL](https://academy.braiins.com/en/braiins-pool/btc-mining-setup/#stratum-v2-endpoints-servers-and-ports) into Bitcoin Core.

In my ideal scenario the miner would find Bitcoin Core via mDNS, but that's a fresh can of worms that I haven't thought through in great detail. So instead a user will have to point to Bitcoin Core in the miner interface (typically a web server running on the device) [0].

To get there we'd have to implement not just the Template Provider role (#29432), but also the Job Declarator role. Additionally we'd have to implement either a (simple) Mining Proxy _or_ finish the Job Distribution Protocol spec, so that the miner can communicate directly with the pool. See the protocol overview: https://github.com/stratum-mining/sv2-spec/blob/main/03-Protocol-Overview.md 

To keep #29432 manageable, and because I don't know if we want to add that much mining functionality, I haven't done this. Instead the user has to install and configure the SRI Job Declarator client role, which also acts as a Proxy role. Probably on the same machine as the node.

We talk to the SRI software using noise encryption, but we _could_ also talk to it with a unix socket. That is, if SRI adds support for that. Regardless of the transport layer, we have to support the sv2 protocol messages.

Implementing Noise encryption (and the sv2 protocol messages) now gives us the flexibility later to take over the Job Declarator client and Proxy role, removing the need to install SRI. At least for simple setups, e.g. a single ASIC (so the proxy doesn't have to worry about how to distribute work).

IIUC the [Job Declarator client role](https://github.com/stratum-mining/sv2-spec/blob/main/06-Job-Declaration-Protocol.md) was designed to keep the Template Provider role easier to implement. To take it over we'd have to implement a few more messages: `AllocateMiningJobToken`, `DeclareMiningJob`, `IdentifyTransactions` and `SubmitSolution`. The `IdentifyTransactions` part is similar to how compact blocks work, where we only send full transactions to the pool if they don't already have them.

While taking over the Job Declarator client role seems like a fairly straight-forward followup, taking over the Mining Proxy role is perhaps less so. Designing and implementing a Job Distribution Protocol, which IIUC avoids the need to implement a Mining Proxy, may be a better path.

If in the end we decide to do neither, then perhaps implementing noise was overkill. But as you point out it's not that complicated.

Any complexity can be hidden from user with node-in-a-box solutions like Umbrel, but it's nice if they can just use Bitcoin Core. Having a separate binary maintained by us, like HWI, adds installation and configuration complexity. For HWI that's unavoidable because we don't want to include the USB driver kitchen sink.

I don't think simplicity is merely nice for amateur home miners with a single S9 space heater; I suspect even small scale "professional" miners are not sys-admins. Running and configuring a seperate job declarator client process making sure it restarts after a crash is non-trivial. Much easier to just point the miner directly to a pool and not bother with custom template creation. So imo we need to lower the barrier as much as (safely) possible.


An additional argument for implementing noise encryption support, completely separate from the above setup complexity, is that it allows making the template provider publicly reachable. I'm still a bit skeptical about that use case, but see discussion here: https://github.com/stratum-mining/sv2-spec/issues/63#issuecomment-1866515958

[0] on a hypothetical stratum v2 aware miner that is, which doesn't exist. Today you have to point the miner to a sv1 - sv2 translator which runs on your computer. This translator points to your job declarator client.

**Comment by Sjors** [comment]
@ryanofsky let's continue the interface design in #30200.

**Comment by Sjors** [comment]
Trivial rebase after #29015.

**Comment by Fi3** [comment]
@Sjors initially the TP was without noise, we added it mainly to have third party TPs (not pool and not miner). About easy of use, if we don't want to implement an sv1 translator in core, the miner still have to deploy a translator proxy. Maybe is easier to just release a translator that is also a JDC (like what I did for demand) rather the implement the JDC in the TP and add custom message to do that.

**Comment by Sjors** [comment]
I'm still hopeful sv1 translators won't be needed at all soon(tm). That certainly seems too complicated to implement in Bitcoin Core.

For the same reason I don't think combining JDC and Translator is a good long term solution, though in the short run it could make the SRI side of things easier to configure.

See also the conceptual discussion in https://github.com/stratum-mining/sv2-spec/discussions/85

**Comment by Sjors** [comment]
The MSan, depends job [failed](https://cirrus-ci.com/task/5872614726434816?logs=ci#L2697) the test which issues the certificate 1 second in the future. I increased the time difference so such failure is less likely.

The fuzzer could also pick 1 second in the past or future for the certificate validity test. But it uses `SetMockTime`, so presumably doesn't have this problem.

**Comment by itornaza** [comment]
After going through https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2134614071, [sv2-spec](https://github.com/stratum-mining/sv2-spec/blob/main/03-Protocol-Overview.md), https://github.com/stratum-mining/sv2-spec/discussions/85 time and time again, I put all the relevant information in [sv2-draft-sketch.pdf](https://github.com/user-attachments/files/16033413/sv2-draft-sketch.pdf) for the Config A case, to make better sense of the trade offs involved.

With the assumption that I understood the protocol correctly, it seems that integrating the Noise Protocol into Bitcoin Core adds much more flexibility and options than overhead https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2110376536. Thus, Noise Protocol integration allows for the following:

1. Implement the Template Provider (TP) role and the Template Distribution Protocol into Bitcoin and point to SRI software for the Job Declarator Client (JDC) and Mining Proxy.
2. Implement the Template Provider (TP), Job Declarator Client (JDC) roles, and Template Distribution and Job Distribution protocols into Bitcoin and point to SRI software for Mining Proxy. This option makes the life of amateur miners easier since they do not need to configure a third party JDC.
3. The Template Provider (TP) can be publicly reachable.

In the longer term, and if it is decided to aim for the incorporation of the Template Provider (TP), Job Declarator Client (JDC) and Mining Proxy into Bitcoin, the Noise Protocol can facilitate all the interim solutions until the rest of the modules are production ready.

**Comment by Fi3** [comment]
@itornaza I agree with you about the implication of integrating noise into the TP. Just want to point out 2 minors issues [in your sketch](https://github.com/user-attachments/files/16033413/sv2-draft-sketch.pdf):
1. Pool and JDS do not need to communicate (and if they do they do not use the job declaration protocol)
2. JDS and TP do not communicate using the template distribution protocol. JDS <-> TP communication is not part of Sv2. For example the JDS on SRI just use bitcoind RPC API.

**Review by KristijanSajenko — APPROVED** [review]

**Comment by itornaza** [comment]
@Fi3 thank you for clarifying it really helped a lot! Would you also think this is a good point to start reviewing the Noise Protocol implementation on this PR compared to it's [specification](https://noiseprotocol.org/noise.html)?

**Comment by Fi3** [comment]
@itornaza  If I remember correctly anything needed to implement sv2 noise is contained in the [sv2 spec](https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md) and the [noise spec](https://noiseprotocol.org/noise.html) are a little bit dispersive. But I implemented it on SRI more then one year ago so take it with a grain of salt.

**Review by winlover32 — APPROVED** [review]

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:26]: nit: Is that comment needed? SV2 spec 4.5.2 is not relevant to message length.

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:70]: styling nit: Consider splitting into multiple lines for clarity and consistency with Serialize above.

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:99]: nit: Should the `plain` message be the @param[in] in the `EncryptWithAdd` method?

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:100]: nit: Should the `ciphertext` message be the @param[out] in the EncryptWithAdd method?

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:237]: styling nit: Consider dropping the `_our` part from the variable's name to be more consistent with the other names and be closer to the Noise protocol naming convention `s` (implicitly our) and `rs` (explicitly remote). Same for the rest of the names below.

File: src/common/sv2_noise.cpp
Comment (itornaza) [inline src/common/sv2_noise.cpp:237]: nit: Consider removing the commented out code. Same to the following line as well.

**Review by itornaza — APPROVED** [review]
Concept ACK a2436955f5bde920e41e03a5db34d477bfeb04a6

As stated in my previous comments, the implementation of the Stratum v2 Noise Protocol is adding more flexibility to Bitcoin than overhead towards the integration of the Stratum protocol Template Provider and Job Declaration Client roles within bitcoind.

I include some really minor nits that came out during a quick code review against the protocol specs and leaving them here to your discretion.

File: src/common/sv2_noise.h
Comment (Sjors) [reply src/common/sv2_noise.h:26]: Looks like this was duplicated from the comment below. Dropped.

File: src/common/sv2_noise.h
Comment (Sjors) [reply src/common/sv2_noise.h:99]: Fixed.

File: src/common/sv2_noise.h
Comment (Sjors) [reply src/common/sv2_noise.h:237]: Done

**Comment by Sjors** [comment]
Rebased after silent merge conflict with #29625 which moved `xoroshiro128plusplus.h`. Addressed comments by @itornaza.

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:128]: Cpp Core Guidelines explicitly want us to avoid C-Arrays [SL.Con.1](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon1-prefer-using-stl-array-or-vector-instead-of-a-c-array) and prefer STL's`std::array` which does not degenerate to a pointer and knows its size. Would it be better for us here to use `std::array<uint8_t, HASHLEN> m_key` or it is not justified in our context?

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:171]: Same here

**Review by Brock124590 — APPROVED** [review]

File: src/common/sv2_noise.h
Comment (Brock124590) [reply src/common/sv2_noise.h:128]: Okay

Comment (Sjors) [reply src/common/sv2_noise.h:128]: @itornaza thanks, I'll look into your patch. Another guiding principle would be to stay close to `bip324.h` in terms of coding style. I haven't compared with that yet.

**Comment by Sjors** [comment]
I can't reproduce the CI failure, but hopefully this fixes it: https://github.com/bitcoin/bitcoin/compare/a19acfccee6f118fad2fb9d090157e5c55cd8465..e73740fde107f79eeee8e55c6bed156713abe8a5

See https://github.com/bitcoin/bitcoin/commit/2d28478d0791689070bd23df5b5640e9dae6f786#r145179932

**Review by itornaza — APPROVED** [review]
Code review ACK e73740fde107f79eeee8e55c6bed156713abe8a5

This pr closely follows the intersection of the original [Noise Protocol paper](https://noiseprotocol.org/noise.pdf) and the [Protocol Security section](https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md) of the Stratum V2 Specification to provide an implementation of the Noise Protocol on Bitcoin Core. In that way we can now step by step integrate the Stratum protocol Template Provider and Job Declaration Client roles into bitcoind.

All my concerns at the present time were addressed and explained in https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2197273180, https://github.com/bitcoin/bitcoin/pull/29346#issuecomment-2198494395 and in my patch's thread https://github.com/bitcoin/bitcoin/commit/2d28478d0791689070bd23df5b5640e9dae6f786 by @Sjors and @Fi3.

I also built the commit again, run all unit, functional and extended tests and all of them pass. I also run the unit tests with the help of lldb to closely follow the program's flow and monitor how the variables behave. I intend to do more testing and review of the sv2 series as I learn more and get familiar with it.

**Comment by Sjors** [comment]
Rebase for CMake (only).

**Comment by Sjors** [comment]
Rebased after #26619.

**Comment by Sjors** [comment]
The first two commits of https://github.com/Sjors/bitcoin/pull/62 could be included here to put it behind a `-DWITH_SV2=ON` feature flag. 

I'll include that in a future push once it had some review. It shouldn't be a blocker for this PR, because the classes introduced here are not used yet.

File: src/common/sv2_noise.cpp
Comment (Shourya742) [inline src/common/sv2_noise.cpp:486]:
```suggestion
    size_t chunk_size = NOISE_MAX_CHUNK_SIZE - Poly1305::TAGLEN;
    size_t num_chunks = (msg_len + chunk_size - 1) / chunk_size;
    return msg_len + (num_chunks * Poly1305::TAGLEN);

```

**Review by bit-aloo — CHANGES_REQUESTED** [review]
utACK. This looks good, just a few nits. It closely aligns with the Noise NX standards and our SRI noise implementation. I will test it soon and share detailed findings.

File: src/common/sv2_noise.h
Comment (itornaza) [inline src/common/sv2_noise.h:151]: If you consider updating for other reasons, could you format this in the same way as PROTOCOL_NAME_HASH?

**Review by itornaza — APPROVED** [review]
re ACK c8d9d84e21ba091be6be83261b31684ac693d016

Run all tests locally once again and all of them pass including the extended.

**Comment by Sjors** [comment]
Added a CMake build flag `-DWITH_SV2` that's `OFF` by default.

Also rebased and addressed comments.

File: src/common/sv2_noise.cpp
Comment (Sjors) [reply src/common/sv2_noise.cpp]: Taken

**Comment by Sjors** [comment]
The macOS 14 native failure is spurious, see #30922.

**Review by stratospher — COMMENTED** [review]
ACK b588ff8. went through the noise protocol spec. also sharing a [pictorial represantion of the NX handshake](https://github.com/stratospher/blogosphere/blob/main/noise.pdf) if it's useful to other reviewers.

File: src/logging.h
Comment (Sjors) [reply src/logging.h:74]: Not sure, maybe 29 existed in an earlier rebase. It might just change it to 29.

File: src/sv2/noise.h
Comment (Sjors) [reply src/sv2/noise.h:150]: It should match the spec. I'm guessing the comment is wrong, because otherwise it wouldn't work against SRI, but will check.

File: src/sv2/noise.h
Comment (Sjors) [reply src/sv2/noise.h:226]: It's only used in tests, but I think we should review the code in both directions under the assumption that they'll be used in production.

**Comment by fanquake** [comment]
> It's ready for review, but not for merge:

> > The parent PR may need more conceptual review

This should probably be a draft then. Re more conceptual review, I think it's been made (fairly?) clear by multiple contributors (and from the progress of the sidecar PRs from #29432) which approach is favoured, and it doesn't seem to be the one involving this PR, and implementing all the SV2 logic inside Core.

**Comment by Sjors** [comment]
I moved this to https://github.com/Sjors/bitcoin/pull/66 so we can focus on building an interface for external applications to use, and getting multiprocess in a release. I plan to open a tracking issue for that.

The followup PRs #30315 and #29432 will also be moved shortly.

@stratospher I will apply your feedback on the moved PR later.

File: src/sv2/noise.h
Comment (Sjors) [reply src/sv2/noise.h:150]: The spec: https://github.com/stratum-mining/sv2-spec/blob/main/04-Protocol-Security.md#451-handshake-act-1-nx-handshake-part-1---e

Which says `Noise_NX_Secp256k1+EllSwift_ChaChaPoly_SHA256`. I adjusted the second comment.

I also briefly sanity checked the numerical values with some online tools.

File: CMakeLists.txt
Comment (Sjors) [reply CMakeLists.txt:241]: Looks like I dropped this in a recent update. `WITH_SV2` is now on by default and not turned off by the fuzzer CI. I think...

File: src/test/fuzz/sv2_noise.cpp
Comment (Sjors) [reply src/test/fuzz/sv2_noise.cpp:89]: It's an edge case, but `valid_from == now` and `valid_to == now` should be allowed.

**Pull request closed without merging**
